### PR TITLE
Feature/relationships

### PR DESCRIPTION
**Pull request by nikita03565**
closes issue #15 

- added query to get relationships. supports filtering by node labels and relationship type
- saving relationships data in a separate state variable. Thought about saving only relationships data and displaying nodes using start/end points but generally we may have nodes that are disconnected and it is helpful to be able to display them
- added color picker for relationships
- made almost everything to display tooltips for relationships but haven't figured out how to apply this to polyline :disappointed: 
- added js destructuring assignments at some places to make code a bit better

Example with a fragment of Saint Petersburg:
image

### CODE REVIEW

File: src/components/Map.js [hunk 1]
```diff
@@ -183,6 +210,20 @@ export class UnconnectedMap extends Component {
 		// this.leafletPolylineLayers[ukey].setConfig??({ color });
 	}
 
+	updateRelationsLayer(data, relationshipData, color, ukey) {
+		// todo check if the layer has changed before rerendering it
+		// this.leafletLayers[ukey].clearLayers();
+		let rgbColor = `rgb(${color.r}, ${color.g}, ${color.b})`;
+		
+		let polylineData = relationshipData.map(part => [part.start, part.end]); // array of arrays does the trick
```
Comment (nikita03565) [inline src/components/Map.js:218]: where i found out about multipolylines from: https://react-leaflet.js.org/docs/example-vector-layers/

File: src/components/Map.js [hunk 2]
```diff
@@ -62,6 +65,19 @@ export class UnconnectedMap extends Component {
...
 				this.updatePolylineLayer(layer.data, layer.color, layer.ukey);
+			} else if (layer.rendering === RENDERING_RELATIONS) {
+				ukeyRelationshipArray.push(layer.ukey);
+				if (this.leafletRelationshipLayers[layer.ukey]) {
+					// todo find a way of updating the polyline layer instead of delete & recreate
+					this.map.removeLayer(this.leafletRelationshipLayers[layer.ukey]);
+				}
+				this.updateRelationsLayer(layer.data, layer.relationshipData, layer.relationshipColor, layer.ukey);
+				// have to use separate marker layer
+				ukeyMarkerArray.push(layer.ukey);
```
Comment (nikita03565) [inline src/components/Map.js:76]: that's not great cos i copy-pasted this part
i'm not even sure that this is the right thing to do but if it is, from code perspective i'd like move this to a small function

File: src/components/layers/Layer.js [hunk 3]
```diff
@@ -175,49 +210,92 @@ export class UnconnectedLayer extends Component {
 		return query;
 	};
 
+	getNodesQuery() {
```
Comment (nikita03565) [inline src/components/layers/Layer.js:213]: just extracted it to smaller function

File: src/components/layers/Layer.js [hunk 4]
```diff
@@ -523,31 +633,45 @@ export class UnconnectedLayer extends Component {
 		/*If layerType==latlon, then we display the elements to choose
            node labels and properties to be used.
          */
-		if (this.state.layerType !== LAYER_TYPE_LATLON)
+		const { rendering, layerType, nodes, nodeLabel, relationships, relationshipLabel,
```
Comment (nikita03565) [inline src/components/layers/Layer.js:636]: couldn't stop myself from doing this. destructuring  is such a great feature, wanted to use it everywhere but it would've introduced way too much unrelated changes

File: src/components/layers/Layer.js [hunk 5]
```diff
@@ -556,37 +680,53 @@ export class UnconnectedLayer extends Component {
...
 						isMulti={false}
-						defaultValue={this.state.tooltipProperty}
+						defaultValue={tooltipProperty}
 						name="tooltipProperty"
 					/>
 				</Form.Group>
 
+				<Form.Group 
+					controlId="formRelationshipTooltipProperty" 
+					hidden={true}  // TODO figure out how to display tooltip at polyline
```
Comment (nikita03565) [inline src/components/layers/Layer.js:708]: not happy about this at all

File: src/services/neo4jService.js [hunk 6]
```diff
@@ -53,7 +53,7 @@ export default {
     let res = [];
     const session = driver.session();
     await session
-      .run(`CALL db.labels() YIELD label RETURN label ORDER BY label`)
+      .run('CALL db.labels() YIELD label RETURN label ORDER BY label')
```
Comment (nikita03565) [inline src/services/neo4jService.js:56]: this is not a template string so no need to use backticks. sorry, python linters spoiled me

File: src/services/neo4jService.js [hunk 7]
```diff
@@ -71,6 +71,19 @@ export default {
     return res;
   },
 
+  getRelationshipLabels: async function (driver) {
+    if (driver == null) return [];
```
Comment (nikita03565) [inline src/services/neo4jService.js:75]: `== null` covers both `=== null` and `=== undefined`

File: src/services/neo4jService.js [hunk 8]
```diff
@@ -181,5 +194,40 @@ export default {
         .catch((error) => {
           return {status: "ERROR", result: error};
         });
+        
   },
+  getRelationshipData: async function (driver, query, params) {
+    const session = driver.session();
+    return await session
```
Comment (nikita03565) [inline src/services/neo4jService.js:201]: i'd like to rewrite this and all existing functions without using `.then` and `.catch`
but again this would cause a lot of unrelated changes

**Comment by nikita03565** [comment]
this project requires an automated code formatter and linter. want to introduce Prettier with ESLint but in the next PR. it'd be impossible to review these changes if i had made it here
i really like neomap, it helps me a lot and i want it to be ..healthy and growing

**Review by stellasia — CHANGES_REQUESTED** [review]
Hi @nikita03565 

Thank you very much for your contribution, that's definitely a nice feature that was missing in this project. 

I also agree with you that linter would be nice to have. Be aware that there are many code improvements on the master_dev branch, that have not yet been merged due to a remaining bug we couldn't identify yet (namely: #82 ), but I will a minima cherry-pick the most important commits to move them to master.

Regarding this PR in particular, I am wondering whether it makes sense to totally split node and relationships rendering? That would reduce the number of properties to be defined for each layer, and also allow us to better support the "Advanced" query for relationship (currently not supported with your implementation). In any case, we need to perform two queries to get the nodes and the relationships, so it won't affect performances. What do you think?

File: src/services/neo4jService.js (hunk 8 above)
Comment (nikita03565) [reply src/services/neo4jService.js:201]: i mean just async/await syntax

File: src/components/layers/Layer.js (hunk 5 above)
Comment (nikita03565) [reply src/components/layers/Layer.js:708]: yeah, didn't think of that. will try

File: src/components/layers/Layer.js [hunk 9]
```diff
@@ -175,49 +210,88 @@ export class UnconnectedLayer extends Component {
...
+		if (layerType === LAYER_TYPE_CYPHER)
 			return this.getCypherQuery();
 
-		if (this.state.layerType === LAYER_TYPE_SPATIAL)
+		if (layerType === LAYER_TYPE_SPATIAL)
 			return this.getSpatialQuery();
+		return this.getNodesQuery();		
+	};
 
+	getRelationshipsQuery() {
```
[the remark this answers is not in the document]
Comment (nikita03565) [reply src/components/layers/Layer.js:261]: yes, for now i intended to show lat/lon only and considering your feedback wanted to add support for other types

**Comment by nikita03565** [comment]
@stellasia what bug are you talking about https://github.com/stellasia/neomap/pull/93#discussion_r650367756 ? want to help with removing this blocker

i'm thinking about starting new branch from `master_dev` with all comments addressed. well actually i'm already working in a new branch. don't want to resolve conflicts and fix things that will break after merge

**Comment by stellasia** [comment]
Hi @nikita03565 , thanks for the shootout and help! I think I have fixed this bug this week-end, let me until the end of this week to finish testing and I'll merge the master_dev branch into master, so that we can restart from a clean repo.
